### DISCUSSION TITLE
Installation Example with Ubuntu, Docker, and Nginx Proxy Manager

### DISCUSSION THREAD

**Discussion by ApuRed** [capcom6/android-sms-gateway#50] [Show and tell]
This example is using a VPS Ubuntu 22.xx with dedicated IP and a subdomain pointing to the IP

It is recommended not to use root as the main user, for this we create a user, give it administrator permissions
sudo adduser sms
sudo usermod -aG sudo sms
sudo su - sms

we can continue with the tutorial

1.- Have a Virtual Machine with docker and docker compose
To install you can use these tutorials:
https://www.digitalocean.com/community/tutorials/how-to-install-and-use-docker-on-ubuntu-22-04
https://www.digitalocean.com/community/tutorials/how-to-install-and-use-docker-compose-on-ubuntu-22-04

2.- Create a docker-compose.yaml file (SMS - MySQL - nginx) - you can use this example and adapt it to your liking
[docker-compose.yaml.txt](https://github.com/capcom6/android-sms-gateway/files/14669551/docker-compose.yaml.txt)

3.- create a config.yml file (you can use the administrator guide at https://sms.capcom.me/getting-started/private-server/)
[config.yml.txt](https://github.com/capcom6/android-sms-gateway/files/14669806/config.yml.txt)

4.- Upload these files to the server (create a preferred folder)

5.- enter the server via putty using cd /[dir-create] and execute the command "docker compose up -d"
![Screenshot_8](https://github.com/capcom6/android-sms-gateway/assets/1470677/1e4651a5-ae2b-4af0-8fcf-407ade1e4412)

6.- When finished, enter server-ip:81 via web to enter the nginx administrator and use the data:
Email: admin@example.com
Password: changeme

7.- Change service access and configure a new proxy host
![Screenshot_5](https://github.com/capcom6/android-sms-gateway/assets/1470677/4f9dfbfc-8778-4378-abfd-f6dfc6272c0a)
![Screenshot_6](https://github.com/capcom6/android-sms-gateway/assets/1470677/7c47344c-31d6-4946-aee1-65a31f62186a)

8.- If everything went well you will be able to see that the service is active.

9.- do not forget to configure the URL and Token in the APP
![WhatsApp Image 2024-03-20 at 11 42 52 AM](https://github.com/capcom6/android-sms-gateway/assets/1470677/d35749e2-396b-4197-b74f-5f1cde789c14)

![WhatsApp Image 2024-03-20 at 11 42 52 AM (1)](https://github.com/capcom6/android-sms-gateway/assets/1470677/9a6ca789-d820-463e-bd7d-b1309285c8a8)

10.- We tested our service with postman.
![Screenshot_9](https://github.com/capcom6/android-sms-gateway/assets/1470677/e7807919-e427-4501-b3e2-20edb6b0da18)

11.- and the test sms will arrive where we indicate
![WhatsApp Image 2024-03-20 at 11 44 59 AM](https://github.com/capcom6/android-sms-gateway/assets/1470677/522f4b2e-40bc-4529-ac45-4cd8b8781d60)

I hope it helps you, if you have questions you can read the documentation or comment to add more details.

PS: All rights and thanks to the creator of this service, it is useful for my business.
PS2: The manual is still being modified, your contributions will help achieve a 100% effective and safe manual.

**Comment by capcom6** [comment]
Hello!

Thank you for providing these detailed instructions. Your contribution will undoubtedly benefit others in the community.

I'd like to share a couple of notes regarding your `docker-compose.yml` setup. I suggest avoiding directly opening ports from `backend` and `db` to mitigate potential security risks.

Here are some additional recommendations:

- You can increase the `tasks.hashing.interval_seconds` value in `config.yml` to minimize the load on the database. I think, frequent hashing isn't necessary when the server is hosted in private mode.
- It appears you initiated `docker compose` with root privileges. It's generally best practice to create a separate user with membership in the `docker` group and run `docker compose` as that user.
- In step 7, concerning the proxy host configuration, using the hostname `backend` to refer to the `backend` container might be more convenient. Since Nginx and the backend are running on the same Docker network, this approach simplifies the configuration. If you remove the port mapping for the `backend` container as I recommended earlier, the setup from your screenshot won't work.

Also, I want to thank you for pointing out `jc21/nginx-proxy-manager`. It's an interesting project that was new to me as well.

I'll be shifting this question over to the discussion section for easier access and further engagement.

Thanks again for your valuable input!

**Comment by yahya567** [comment]
Thank you for the comprehensive guide. It is well-constructed and detailed. However, I noticed there is no information on how to obtain the necessary credentials. Could you please provide guidance on where to acquire these credentials? I attempted to use the Authorization generated by the app when connecting to the server, but it did not work. 

I tried executing the POST request to `https://my-domain/api/3rdparty/v1/message `and received the following message:

```
{
    "message": "Unauthorized"
}
```

Thank you very much.

**Reply by capcom6** [reply]
@yahya567, hello!

The credentials are displayed on the Home screen of the app. However, if you are using a Private Server, please ensure that you set the full URL (including path `/api/mobile/v1`) and token in the app settings. Don't forget to restart the app by using the button on the Home screen.

See also https://sms.capcom.me/getting-started/private-server/

**Comment by yahya567** [comment]
Worked. Thank you.

**Comment by KleKlai** [comment]
Before i begin i would like to thank you @ApuRed for the detailed guides and @capcom6 for making the project open source 😊

**Describe the Bug**
When attempting to access the SMS Gateway backend API through xxx.xxx.xxx.xxx:3000/api, instead of the expected Swagger UI, a white screen is displayed. The console shows multiple 404 errors, indicating that the resources (such as CSS and JS files for Swagger UI) are not being found.

**To Reproduce**

1. Deploy SMS Gateway containers using Docker Compose on a Hetzner server running Ubuntu 2024.
2. Ensure all containers are running as verified in Portainer.
3. Attempt to access the API endpoint via http://xxx.xxx.xxx.xxx:3000/api.
4. Observe that a white screen is displayed, and the console logs 404 errors.

**Expected vs Actual Behavior**

1. Expected: Swagger UI should load correctly when accessing the API endpoint, allowing interaction with the API documentation.
2. Actual: The Swagger UI fails to load, resulting in a white screen and multiple 404 errors in the console.
3. Screenshots

The provided screenshots illustrate the running containers and the console errors encountered when attempting to load the Swagger UI.
![Screenshot_20240813_124222](https://github.com/user-attachments/assets/68a0fd32-404c-4c53-ba1f-865813c680b6)
![Screenshot_20240813_123654](https://github.com/user-attachments/assets/ae3c66f9-5938-4276-a228-c96f852ceb88)

**Environment**

- App Version: Likely the latest version as pulled by Docker.
- Mode: Cloud (on Hetzner server)
- Operation: Setting up and testing SMS Gateway backend
- Network Type: Static IP provided by Hetzner

**Additional Context**
It appears that the backend API is not correctly serving the Swagger UI resources, leading to the 404 errors. This could be due to incorrect file paths, misconfigured environment variables, or issues in the Docker configuration.

Did i forgot something to configure? i just follow the guide and the documentation on setting up the private server.

**Reply by capcom6** [reply]
Hello!

It's possible that the issue is due to a "trailing slash problem". Try adding a slash at the end of the path, like this: xxx.xxx.xxx.xxx:3000/api/.

I will try to fix it in the next version.

**Reply by KleKlai** [reply]
hi @capcom6 thank you for your reply. 

I can now access swaggerUI 

![image](https://github.com/user-attachments/assets/02087e77-3896-4a4b-a146-b1c30da1b784)

using the solution you provided about trailing slash problem. Thank you 😁

**Reply by capcom6** [reply]
Thanks for your feedback!

Will be fixed in next server release: https://github.com/capcom6/sms-gateway/pull/84

**Comment by rohit9625** [comment]
@ApuRed Do we necessarily need a domain name? From where did you get the private_token?

**Reply by capcom6** [reply]
Hello!

While it's not strictly necessary, using a domain name is strongly recommended for a Private Server setup. There can be issues with obtaining SSL certificates for IP addresses. For example, Let's Encrypt doesn't issue certificates for IP addresses at all.

Regarding the `private_token`, it's a security measure you set up yourself. It acts as a unique password that you configure on your server and then provide to the Android app in the "Cloud Server" section of the Settings. You choose this token, ensuring it's sufficiently strong to prevent unauthorized access. For more detailed information on setting up a Private Server, please refer to the documentation at https://docs.sms-gate.app/getting-started/private-server/

**Reply by rohit9625** [reply]
Thank you for the reply. I wanted to test the server that's why I didn't add a domain. Would it be possible to test without domain?

**Reply by capcom6** [reply]
The app requires a valid SSL certificate, which can be challenging to obtain for an IP address alone.

For testing purposes, you might find it easier to use a service like [ngrok](https://ngrok.com/docs/getting-started/). Ngrok creates a temporary domain with a valid SSL certificate and proxies all requests to your server instance. This approach allows you to quickly set up a secure connection for testing without the need to configure a permanent domain or SSL certificate.

**Reply by rohit9625** [reply]
Thank you for the reference, @capcom6. I'll definitely try ngrok and will let you know once it's been tested successfully. Thanks a lot.

**Comment by crimson0714AM** [comment]
Hi, I hope someone can help me with this.

I installed the SMS Gateway on an Ubuntu Server using Docker Compose and Portainer, and I pulled the latest image. However, I’m running into an issue where the SMS Gateway backend container is not exposing the published port.

When I check the logs, I see that the backend container is stuck in a restart loop and keeps restarting continuously.

Could someone please help me investigate this? I will attach screenshots/logs for reference.

Thank you in advance.
<img width="1008" height="503" alt="image1" src="https://github.com/user-attachments/assets/c00b0fbf-c414-4938-aa04-e07f34f375f7" />
<img width="1387" height="282" alt="image2" src="https://github.com/user-attachments/assets/d636bbdd-ed3f-4d07-8777-2da715c72baa" />
<img width="1326" height="691" alt="image3" src="https://github.com/user-attachments/assets/3eaaac41-d2db-448a-8e4d-f854e5a5cd2c" />

**Reply by capcom6** [reply]
Hello!

According to the log, you started the server in public mode without specifying FCM credentials.

To fix this, specify the FCM credentials JSON data in the `FCM__CREDENTIALS_JSON` environment variable or the corresponding YAML configuration key. Alternatively, switch the server to private mode by setting `GATEWAY__MODE` to `private`.
